File: mobile/src/lib/protocol.ts
// Espejo TypeScript del protocolo de VoxLFA para la app móvil.
//
// Regla del proyecto: estos tipos coinciden 1:1 con `core/src/protocol/`
// (Rust/serde) y con `desktop/src/lib/types.ts`. Nunca cambies un nombre de
// campo sin actualizar los tres lados. Este archivo NO depende de Tauri.

/** Estados posibles del motor de audio. */
export type EngineState =
  | "stopped"
  | "starting"
  | "running"
  | "stopping"
  | "error";

/** Descripción de un dispositivo de audio del sistema. */
export interface AudioDeviceInfo {
  name: string;
  isDefault: boolean;
}

/** Instantánea del estado y configuración del motor. */
export interface EngineStatus {
  state: EngineState;
  sampleRate: number;
  bufferSize: number;
  latencyMs: number;
  /** Host de audio activo (p. ej. `"alsa"`, `"jack"`). */
  audioHost: string | null;
  inputDevice: string | null;
  outputDevice: string | null;
}

/** Muestra de nivel del audio capturado/procesado y latencia actual. */
export interface LevelSample {
  inputRmsDb: number;
  inputPeakDb: number;
  outputRmsDb: number;
  outputPeakDb: number;
  latencyMs: number;
  capturedAtMs: number;
}

/** Número fijo de bandas logarítmicas del espectro emitido por el motor. */
export const SPECTRUM_BIN_COUNT = 32;

/** Muestra del espectro de la entrada (FFT) emitida en vivo. */
export interface SpectrumSample {
  /** Nivel de cada banda logarítmica en dBFS (longitud fija). */
  binsDb: number[];
  /** Frecuencia de muestreo (Hz) de la captura; define los bordes de banda. */
  sampleRate: number;
  /** Tiempo monotónico (ms) de la captura. */
  capturedAtMs: number;
}

/** Identificador de un preset de la cabina. */
export type PresetId = "dry" | "vozLimpia" | "radio" | "warm";

/** Banda de un ecualizador paramétrico. */
export interface EqBand {
  kind: EqBandKind;
  freqHz: number;
  gainDb: number;
  q: number;
}

/** Tipo de banda del ecualizador. */
export type EqBandKind = "lowShelf" | "peaking" | "highShelf";

/** Parámetros de la puerta de ruido de la cadena. */
export interface NoiseGateParams {
  /** Umbral de cierre en dBFS. */
  thresholdDb: number;
  /** Tiempo de ataque (ms). */
  attackMs: number;
  /** Tiempo de liberación (ms). */
  releaseMs: number;
  /** Tiempo de retención de la apertura (ms). */
  holdMs: number;
  /** Atenuación máxima cuando la puerta está cerrada (dB). */
  rangeDb: number;
}

/** Nota musical raíz para la escala de corrección de tono. */
export type MusicalNote =
  | "c" | "cs" | "d" | "ds" | "e" | "f"
  | "fs" | "g" | "gs" | "a" | "as" | "b";

/** Escala musical para la corrección de tono. */
export type MusicalScale =
  | "chromatic" | "major" | "minorNatural" | "minorHarmonic"
  | "pentatonicMajor" | "pentatonicMinor" | "blues";

/** Parámetros de corrección de tono. */
export interface PitchCorrectionParams {
  scale: MusicalScale;
  root: MusicalNote;
  strength: number;
  mix: number;
}

/** Parámetros de supresión de ruido. */
export interface DenoiseParams {
  /** Mezcla seco/húmedo (0 = sin denoise, 1 = denoise completo). */
  mix: number;
}

/** Parámetros de supresión de feedback adaptativa. */
export interface FeedbackSuppressorParams {
  /** Umbral de detección en dBFS. */
  thresholdDb: number;
  /** Factor de calidad de los filtros notch. */
  q: number;
}

/** Modo del delay. */
export type DelayMode = "digital" | "analog" | "tape" | "slapback";

/** Modo del reverb. */
export type ReverbMode = "plate" | "hall" | "room";

/** Parámetros de delay. */
export interface DelayParams {
  mode: DelayMode;
  timeMs: number;
  feedback: number;
  mix: number;
  preDelayMs: number;
  lowCutHz: number;
  highCutHz: number;
  tempoBpm: number;
  syncEnabled: boolean;
  duckAmount: number;
}

/** Parámetros de reverb. */
export interface ReverbParams {
  mode: ReverbMode;
  roomSize: number;
  damping: number;
  wet: number;
  preDelayMs: number;
  highCutHz: number;
  lowCutHz: number;
}

/** Estado de un módulo dentro de la cadena activa. */
export interface DspLinkState {
  name: string;
  enabled: boolean;
  bypass: boolean;
  /** Bandas actuales del EQ si este módulo es el ecualizador; si no, `null`. */
  eqBands: EqBand[] | null;
  /** Parámetros actuales de la puerta si este módulo es la puerta; si no, `null`. */
  gateParams: NoiseGateParams | null;
  /** Parámetros de denoise si este módulo es denoise; si no, `null`. */
  denoiseParams: DenoiseParams | null;
  /** Parámetros de feedback si este módulo es feedback; si no, `null`. */
  feedbackParams: FeedbackSuppressorParams | null;
  /** Parámetros de pitch correction si este módulo es pitch correction; si no, `null`. */
  pitchCorrectionParams: PitchCorrectionParams | null;
  /** Parámetros de delay si este módulo es delay; si no, `null`. */
  delayParams: DelayParams | null;
  /** Parámetros de reverb si este módulo es reverb; si no, `null`. */
  reverbParams: ReverbParams | null;
}

/** Estado completo de la cadena DSP activa. */
export interface DspState {
  preset: PresetId;
  globalBypass: boolean;
  links: DspLinkState[];
}

/** Listado de dispositivos de entrada/salida. */
export interface DeviceList {
  inputs: AudioDeviceInfo[];
  outputs: AudioDeviceInfo[];
}

/** Métricas de la voz sobre la ventana deslizante de análisis (dBFS). */
export interface VoiceMetrics {
  rmsDb: number;
  peakDb: number;
  dynamicRangeDb: number;
  crestDb: number;
  brightness: number;
  resonanceScore: number;
  fatigueScore: number;
  windowMs: number;
}

/** Área de la voz que motiva una sugerencia. */
export type SuggestionKind =
  | "timbre"
  | "dynamics"
  | "fatigue"
  | "resonance"
  | "aiAdvisor";

/** Acción confirmable que acompaña a una sugerencia. */
export type SuggestionAction =
  | { type: "none" }
  | { type: "applyPreset"; preset: PresetId }
  | { type: "setEqBand"; bandIndex: number; gainDb: number }
  | { type: "setDenoise"; mix: number }
  | { type: "setFeedback"; thresholdDb: number; q: number }
  | { type: "setPitchCorrection"; strength: number; mix: number }
  | { type: "setNoiseGate"; thresholdDb: number; rangeDb: number }
  | { type: "setDelay"; timeMs: number; mix: number }
  | { type: "setReverb"; wet: number; roomSize: number };

/** Sugerencia generada por el asistente para la voz actual. */
export interface Suggestion {
  id: number;
  kind: SuggestionKind;
  severity: number;
  message: string;
  action: SuggestionAction;
}

/** Muestra de análisis emitida por el motor (métricas + sugerencias). */
export interface AnalysisSample {
  metrics: VoiceMetrics;
  suggestions: Suggestion[];
  capturedAtMs: number;
}

/** Resumen acumulado de la sesión de voz en curso. */
export interface SessionSummary {
  startedAtMs: number;
  durationMs: number;
  avgRmsDb: number;
  peakDb: number;
  dynamicRangeDb: number;
  avgBrightness: number;
  fatigueScore: number;
  loudTimeMs: number;
  suggestionsCount: number;
}

/**
 * Comando de control enviado por el móvil al escritorio (tag `type`,
 * campos camelCase). Espejo de `core/src/protocol/control.rs`.
 *
 * `start` NO se incluye a propósito: arrancar el motor solo se permite desde
 * la cabina de escritorio.
 */
export type ControlCommand =
  | { type: "stop" }
  | { type: "setPreset"; preset: PresetId }
  | { type: "setGlobalBypass"; bypass: boolean }
  | { type: "setLinkBypass"; link: string; bypass: boolean }
  | { type: "setEqBand"; bandIndex: number; gainDb: number }
  | { type: "setNoiseGate"; params: NoiseGateParams }
  | { type: "setDenoise"; params: DenoiseParams }
  | { type: "setFeedback"; params: FeedbackSuppressorParams }
  | { type: "setPitchCorrection"; params: PitchCorrectionParams }
  | { type: "setDelay"; params: DelayParams }
  | { type: "setReverb"; params: ReverbParams };

/**
 * Evento emitido por el motor por el WebSocket (tag `type`, campos camelCase).
 */
export type EngineEvent =
  | (EngineStatus & { type: "status" })
  | (LevelSample & { type: "level" })
  | (SpectrumSample & { type: "spectrum" })
  | (DeviceList & { type: "devices" })
  | (DspState & { type: "dsp" })
  | (AnalysisSample & { type: "analysis" })
  | { type: "warning"; message: string };

/** Guard de tipos para mensajes recibidos del WebSocket. */
export function isEngineEvent(raw: unknown): raw is EngineEvent {
  if (typeof raw !== "object" || raw === null) return false;
  const event = raw as Record<string, unknown>;
  if (typeof event.type !== "string") return false;
  switch (event.type) {
    case "status":
      return typeof event.sampleRate === "number" && typeof event.latencyMs === "number";
    case "level":
      return (
        typeof event.inputRmsDb === "number" &&
        typeof event.inputPeakDb === "number" &&
        typeof event.outputRmsDb === "number" &&
        typeof event.outputPeakDb === "number"
      );
    case "spectrum":
      return (
        Array.isArray(event.binsDb) &&
        event.binsDb.every((value) => typeof value === "number") &&
        typeof event.sampleRate === "number" &&
        typeof event.capturedAtMs === "number"
      );
    case "devices":
      return Array.isArray(event.inputs) && Array.isArray(event.outputs);
    case "dsp":
      return (
        typeof event.preset === "string" &&
        typeof event.globalBypass === "boolean" &&
        Array.isArray(event.links)
      );
    case "analysis":
      return (
        typeof event.capturedAtMs === "number" &&
        typeof event.metrics === "object" &&
        event.metrics !== null &&
        Array.isArray(event.suggestions)
      );
    case "warning":
      return typeof event.message === "string";
    default:
      return false;
  }
}

File: core/tests/protocol.rs
//! Tests de integración del protocolo: el contrato JSON que se usa entre el
//! core, la UI de escritorio y la app móvil.

use voxlfa_core::protocol::{
    AudioDeviceInfo, ControlCommand, EngineEvent, EngineState, EngineStatus, LevelSample,
};
use voxlfa_core::Result;

#[test]
fn engine_event_level_serializes_with_camel_case_and_tag() {
    let event = EngineEvent::Level(LevelSample {
        input_rms_db: -24.5,
        input_peak_db: -12.0,
        output_rms_db: -26.0,
        output_peak_db: -13.0,
        latency_ms: 12.3,
        captured_at_ms: 123456,
    });

    let json = serde_json::to_string(&event).unwrap();
    // El discriminante del enum debe ser `type` y los campos en camelCase.
    assert!(json.contains("\"type\":\"level\""));
    assert!(json.contains("\"inputRmsDb\":"));
    assert!(json.contains("\"inputPeakDb\":"));
    assert!(json.contains("\"outputRmsDb\":"));
    assert!(json.contains("\"outputPeakDb\":"));
    assert!(json.contains("\"latencyMs\":"));
    assert!(json.contains("\"capturedAtMs\":"));
}

#[test]
fn engine_event_status_round_trips() {
    let event = EngineEvent::Status(EngineStatus {
        state: EngineState::Running,
        sample_rate: 48_000,
        buffer_size: 256,
        latency_ms: 10.0,
        audio_host: Some("alsa".into()),
        input_device: Some("Mic".into()),
        output_device: None,
    });

    let json = serde_json::to_string(&event).unwrap();
    assert!(json.contains("\"type\":\"status\""));
    assert!(json.contains("\"sampleRate\":48000"));

    let decoded: EngineEvent = serde_json::from_str(&json).unwrap();
    match decoded {
        EngineEvent::Status(status) => {
            assert_eq!(status.state, EngineState::Running);
            assert_eq!(status.sample_rate, 48_000);
        }
        other => panic!("esperaba Status, obtuve {other:?}"),
    }
}

#[test]
fn engine_event_devices_round_trips() {
    let event = EngineEvent::Devices {
        inputs: vec![AudioDeviceInfo {
            name: "Built-in Microphone".into(),
            is_default: true,
        }],
        outputs: vec![AudioDeviceInfo {
            name: "Speakers".into(),
            is_default: true,
        }],
    };

    let json = serde_json::to_string(&event).unwrap();
    let decoded: EngineEvent = serde_json::from_str(&json).unwrap();
    match decoded {
        EngineEvent::Devices { inputs, outputs } => {
            assert_eq!(inputs.len(), 1);
            assert_eq!(outputs[0].name, "Speakers");
        }
        other => panic!("esperaba Devices, obtuve {other:?}"),
    }
}

#[test]
fn control_command_start_uses_camel_case() {
    let command = ControlCommand::Start {
        input_device: Some("Mic A".into()),
        output_device: None,
    };

    let json = serde_json::to_string(&command).unwrap();
    assert!(json.contains("\"type\":\"start\""));
    assert!(json.contains("\"inputDevice\":\"Mic A\""));
    assert!(json.contains("\"outputDevice\":null"));

    let decoded: ControlCommand = serde_json::from_str(&json).unwrap();
    match decoded {
        ControlCommand::Start {
            input_device,
            output_device,
        } => {
            assert_eq!(input_device.as_deref(), Some("Mic A"));
            assert!(output_device.is_none());
        }
        other => panic!("esperaba Start, obtuve {other:?}"),
    }
}

#[test]
fn control_command_remote_variants_use_camel_case_and_round_trip() {
    use voxlfa_core::protocol::PresetId;

    let commands = vec![
        ControlCommand::Stop,
        ControlCommand::SetPreset {
            preset: PresetId::Warm,
        },
        ControlCommand::SetGlobalBypass { bypass: true },
        ControlCommand::SetLinkBypass {
            link: "eq".into(),
            bypass: true,
        },
        ControlCommand::SetEqBand {
            band_index: 2,
            gain_db: -4.5,
        },
    ];

    for command in commands {
        let json = serde_json::to_string(&command).unwrap();
        let decoded: ControlCommand = serde_json::from_str(&json).unwrap();
        match (command, decoded) {
            (ControlCommand::Stop, ControlCommand::Stop) => {
                assert_eq!(json, r#"{"type":"stop"}"#);
            }
            (ControlCommand::SetPreset { preset }, ControlCommand::SetPreset { preset: p }) => {
                assert_eq!(preset, p);
                assert!(json.contains("\"type\":\"setPreset\""));
                assert!(json.contains("\"preset\":\"warm\""));
            }
            (
                ControlCommand::SetGlobalBypass { bypass },
                ControlCommand::SetGlobalBypass { bypass: b },
            ) => {
                assert_eq!(bypass, b);
                assert!(json.contains("\"type\":\"setGlobalBypass\""));
                assert!(json.contains("\"bypass\":true"));
            }
            (
                ControlCommand::SetLinkBypass { link, bypass },
                ControlCommand::SetLinkBypass { link: l, bypass: b },
            ) => {
                assert_eq!(link, l);
                assert_eq!(bypass, b);
                assert!(json.contains("\"type\":\"setLinkBypass\""));
                assert!(json.contains("\"link\":\"eq\""));
            }
            (
                ControlCommand::SetEqBand {
                    band_index,
                    gain_db,
                },
                ControlCommand::SetEqBand {
                    band_index: i,
                    gain_db: g,
                },
            ) => {
                assert_eq!(band_index, i);
                assert_eq!(gain_db, g);
                assert!(json.contains("\"type\":\"setEqBand\""));
                assert!(json.contains("\"bandIndex\":2"));
                assert!(json.contains("\"gainDb\":-4.5"));
            }
            other => panic!("el comando no preservó su variante: {other:?}"),
        }
    }
}

#[test]
fn unknown_control_command_type_is_rejected() {
    let invalid = r#"{"type":"fly"}"#;
    let result: Result<ControlCommand> = serde_json::from_str(invalid).map_err(Into::into);
    assert!(result.is_err());
}

#[test]
fn control_command_new_dsp_variants_round_trip() {
    use voxlfa_core::protocol::{
        DenoiseParams, FeedbackSuppressorParams, MusicalNote, MusicalScale, NoiseGateParams,
        PitchCorrectionParams,
    };

    let commands = vec![
        ControlCommand::SetNoiseGate {
            params: NoiseGateParams {
                threshold_db: -40.0,
                attack_ms: 1.0,
                release_ms: 100.0,
                hold_ms: 200.0,
                range_db: 30.0,
            },
        },
        ControlCommand::SetDenoise {
            params: DenoiseParams { mix: 0.75 },
        },
        ControlCommand::SetFeedback {
            params: FeedbackSuppressorParams {
                threshold_db: -30.0,
                q: 10.0,
            },
        },
        ControlCommand::SetPitchCorrection {
            params: PitchCorrectionParams {
                scale: MusicalScale::Major,
                root: MusicalNote::A,
                strength: 0.5,
                mix: 0.8,
            },
        },
    ];

    for command in commands {
        let json = serde_json::to_string(&command).unwrap();
        let decoded: ControlCommand = serde_json::from_str(&json).unwrap();
        match (&command, &decoded) {
            (ControlCommand::SetNoiseGate { .. }, ControlCommand::SetNoiseGate { .. }) => {
                assert!(json.contains("\"type\":\"setNoiseGate\""));
            }
            (ControlCommand::SetDenoise { .. }, ControlCommand::SetDenoise { .. }) => {
                assert!(json.contains("\"type\":\"setDenoise\""));
            }
            (ControlCommand::SetFeedback { .. }, ControlCommand::SetFeedback { .. }) => {
                assert!(json.contains("\"type\":\"setFeedback\""));
            }
            (
                ControlCommand::SetPitchCorrection { .. },
                ControlCommand::SetPitchCorrection { .. },
            ) => {
                assert!(json.contains("\"type\":\"setPitchCorrection\""));
            }
            other => panic!("el comando no preservó su variante: {other:?}"),
        }
    }
}

#[test]
fn engine_event_dsp_serializes_with_type_tag() {
    use voxlfa_core::protocol::{DspLinkState, DspState, EqBand, EqBandKind, PresetId};

    let event = EngineEvent::Dsp(DspState {
        preset: PresetId::VozLimpia,
        global_bypass: false,
        links: vec![DspLinkState {
            name: "eq".into(),
            enabled: true,
            bypass: false,
            eq_bands: Some(vec![EqBand {
                kind: EqBandKind::Peaking,
                freq_hz: 3000.0,
                gain_db: 2.0,
                q: 1.5,
            }]),
            gate_params: None,
            denoise_params: None,
            feedback_params: None,
            pitch_correction_params: None,
            delay_params: None,
            reverb_params: None,
        }],
    });

    let json = serde_json::to_string(&event).unwrap();
    assert!(json.contains("\"type\":\"dsp\""));
    assert!(json.contains("\"preset\":\"vozLimpia\""));
    assert!(json.contains("\"globalBypass\":false"));
    assert!(json.contains("\"links\":"));
    // Las bandas del EQ viajan en camelCase dentro del estado del eslabón.
    assert!(json.contains("\"eqBands\":"));
    assert!(json.contains("\"freqHz\":3000.0"));
    assert!(json.contains("\"gainDb\":2.0"));

    let decoded: EngineEvent = serde_json::from_str(&json).unwrap();
    match decoded {
        EngineEvent::Dsp(state) => {
            assert_eq!(state.preset, PresetId::VozLimpia);
            assert_eq!(state.links[0].name, "eq");
            assert_eq!(state.links[0].eq_bands.as_ref().unwrap().len(), 1);
        }
        other => panic!("esperaba Dsp, obtuve {other:?}"),
    }
}

#[test]
fn unknown_event_type_is_rejected() {
    let invalid = r#"{"type":"nope","rmsDb":-1.0}"#;
    let result: Result<EngineEvent> = serde_json::from_str(invalid).map_err(Into::into);
    assert!(result.is_err());
}
